src/support/global.c

```diff
@@ -97,45 +147,43 @@ __global_calibrate_ticks(void)
     {
         struct timespec start, stop;
         double ratio;
+        uint64_t count;
         uint64_t diff_nsec, diff_tsc, min_nsec, min_tsc;
         uint64_t tries, tsc_start, tsc_stop;
         volatile uint64_t i;
 
+        /* Quickly pre-calibrate to find approximate clock speed */
+        count = WT_MILLION / 10;
+        CLOCK_CALIBRATE_ONE();
+        ratio = (double)diff_tsc / (double)diff_nsec;
+        if (ratio <= DBL_EPSILON)
+            return;
+
+        /* Calculate the number of iterations per 1 ms worth of time */
+        count = (uint64_t)((double)count * CLOCK_CALIBRATE_ONE_MS * 1e6 / diff_nsec);
+        if (count < WT_MILLION)
+            /* Running less than this could lead to drop of precision. */
+            count = WT_MILLION;
+
         /*
          * Run this calibration loop a few times to make sure we get a reading that does not have a
          * potential scheduling shift in it. The inner loop is CPU intensive but a scheduling change
          * in the middle could throw off calculations. Take the minimum amount of time and compute
          * the ratio.
          */
         min_nsec = min_tsc = UINT64_MAX;
-        for (tries = 0; tries < 3; ++tries) {
-            /* This needs to be CPU intensive and large enough. */
-            __wt_epoch(NULL, &start);
-            tsc_start = __wt_rdtsc();
-            for (i = 0; i < 100 * WT_MILLION; i++)
-                ;
-            tsc_stop = __wt_rdtsc();
-            __wt_epoch(NULL, &stop);
-            diff_nsec = WT_TIMEDIFF_NS(stop, start);
-            diff_tsc = tsc_stop - tsc_start;
-
-            /* If the clock didn't tick over, we don't have a sample. */
-            if (diff_nsec == 0 || diff_tsc == 0)
-                continue;
+        for (tries = 0; tries < CLOCK_CALIBRATE_TRIES; ++tries) {
+            CLOCK_CALIBRATE_ONE();
             min_nsec = WT_MIN(min_nsec, diff_nsec);
             min_tsc = WT_MIN(min_tsc, diff_tsc);
         }
+        WT_ASSERT(NULL, min_nsec != UINT64_MAX && min_nsec != 0);
+        WT_ASSERT(NULL, min_tsc != UINT64_MAX && min_tsc != 0);
 
-        /*
-         * Only use rdtsc if we got a good reading. One reason this might fail is that the system's
-         * clock granularity is not fine-grained enough.
-         */
-        if (min_nsec != UINT64_MAX) {
```

Comment:
Note that this check is no longer needed because the test for sensible measurements is now done inside of the "tries" loop. If measurements are bad, we don't come here.